pkg/controllers/management/nodetemplate/nodetemplate_handler.go

```diff
@@ -0,0 +1,281 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/rancher/rancher/pkg/controllers/management/rbac"
+	"github.com/rancher/rancher/pkg/namespace"
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	ownerBindingsAnno = "ownerBindingsCreated"
+)
+
+type nodeTemplateController struct {
+	roleLister v3.GlobalRoleLister
+	roleClient v3.GlobalRoleInterface
+	rbLister   v3.GlobalRoleBindingLister
+	rbClient   v3.GlobalRoleBindingInterface
+	ntClient   v3.NodeTemplateInterface
+	ntLister   v3.NodeTemplateLister
+	npLister   v3.NodePoolLister
+	npClient   v3.NodePoolInterface
+	nsLister   v1.NamespaceLister
+	nsClient   v1.NamespaceInterface
+	mgmtCtx    *config.ManagementContext
+}
+
+func Register(ctx context.Context, mgmt *config.ManagementContext) {
+	nt := nodeTemplateController{
+		roleLister: mgmt.Management.GlobalRoles("").Controller().Lister(),
+		roleClient: mgmt.Management.GlobalRoles(""),
+		rbLister:   mgmt.Management.GlobalRoleBindings("").Controller().Lister(),
+		rbClient:   mgmt.Management.GlobalRoleBindings(""),
+		ntClient:   mgmt.Management.NodeTemplates(""),
+		ntLister:   mgmt.Management.NodeTemplates("").Controller().Lister(),
+		npLister:   mgmt.Management.NodePools("").Controller().Lister(),
+		npClient:   mgmt.Management.NodePools(""),
+		nsLister:   mgmt.Core.Namespaces("").Controller().Lister(),
+		nsClient:   mgmt.Core.Namespaces(""),
+		mgmtCtx:    mgmt,
+	}
+
+	mgmt.Management.NodeTemplates("").Controller().AddHandler(ctx, "nt-grb-handler", nt.sync)
+}
+
+func (nt *nodeTemplateController) sync(key string, nodeTemplate *v3.NodeTemplate) (runtime.Object, error) {
+	if nodeTemplate == nil || nodeTemplate.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	metaAccessor, err := meta.Accessor(nodeTemplate)
+	if err != nil {
+		return nodeTemplate, err
+	}
+
+	creatorID, ok := metaAccessor.GetAnnotations()[rbac.CreatorIDAnn]
+	if !ok {
+		return nodeTemplate, fmt.Errorf("clusterTemplate %v has no creatorId annotation", metaAccessor.GetName())
+	}
+
+	// dynamic client must be used to avoid overwriting dynamic data when updating/cloning node template
+	restConfig := nt.mgmtCtx.RESTConfig
+	dynamicClient, err := dynamic.NewForConfig(&restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	s := schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodetemplates",
+	}
+	ntDynamicClient := dynamicClient.Resource(s)
+
+	migratedNTName := fmt.Sprintf("nt-%s-%s", nodeTemplate.Namespace, nodeTemplate.Name)
+	if nodeTemplate.Namespace != namespace.NodeTemplateGlobalNamespace {
```

Comment:
maybe since this only gonna technically happen once it could move to another function so the sync handler reads a little cleaner?

Reply:
good point, moved